devdoc/sm_requirements.md

```diff
@@ -198,6 +202,23 @@ MOCKABLE_FUNCTION(, int, sm_close_begin, SM_HANDLE, sm);
 
 **SRS_SM_02_071: [** If there are any failures then `sm_close_begin` shall fail and return `SM_ERROR`. **]**
 
+### sm_close_begin_with_cb
+```c
+MOCKABLE_FUNCTION(, SM_RESULT, sm_close_begin_with_cb, SM_HANDLE, sm, SM_CLOSING_COMPLETE_CALLBACK, callback, void*, callback_context);
+```
+
+`sm_close_begin_with_cb` asks from `sm` permission to exit `SM_OPENED` state (or one of its derived state) and return to `SM_CREATED`. `sm_close_begin_with_cb` invokes the `callback` function with `callback_context` before waiting for pending calls to become 0.
+
+**S_RS_SM_28_001: [** If `sm` is `NULL` then `sm_close_begin_with_cb` shall fail and return `SM_ERROR`. **]**
+
+**S_RS_SM_28_002: [** If `callback` is `NULL` then `sm_close_begin_with_cb` shall fail and return `SM_ERROR`. **]**
+
+**S_RS_SM_28_003: [** `sm_close_begin_with_cb` shall behave as if `sm_close_begin` was called except that `callback` will be invoked with `callback_context` as arguments before waiting for pending calls to become 0. **]**
```

Comment:
>`sm_close_begin_with_cb` shall behave as if `sm_close_begin` was called except that `cal [](http://example.com/codeflow?start=22&length=88)

consider re-writing this like:

first spec: ...shall behave as if sm_close_begin was called except as follows:
second spec: after switching the state to `SM_OPENED_DRAINING_TO_CLOSE` ...shall invoke the callback...

I just think it's 2 specs, first is that it does the same as the normal close, second is that it calls the callback at a specific point in time.